File: python/game/Game.py
import copy
import os
import time


from game import Config
from game.graphics.GUI import GUI
from game.player.Player import Player
from game.state.StateManager import StateManager
from sfml.sf import Clock

from game.environment.Tilemap import Tilemap


class Game:
    def __init__(self):
        self.map = Tilemap("contested-4v4.json")
        self.state_manager = StateManager()

        self.units = []
        self.players = []

        self.update_next = None
        self.render_next = None
        self.apm_next = None
        self.stats_next = None

        self.update_interval = 1000000 / 600
        self.render_interval = 1000000 / 60
        self.apm_interval = 1000000 / 600

        self.current_ups = 0
        self.current_fps = 0

        self.update_delta = 0
        self.render_delta = 0

        self.ticks = 0

        self.doDisplay = True
        self.doCaptionConsole = True
        self.doCaptionWindow = False


        # Create GUI
        self.gui = GUI(self)

        self.clock = Clock()

    def add_player(self):
        player = Player(self, len(self.players))
        self.players.append(player)

        self.spawn_player(player)

        return player

    def spawn_player(self, player):

        spawn_point_index = self.map.spawn_tiles[player.id]
        spawn_tile = self.map.tiles[spawn_point_index]


        builder = player.spawn(spawn_tile)

        if Config.MECHANICS_TOWN_HALL_ON_START:
            builder.build(0)


    def loop(self):
        now = self.clock.elapsed_time.microseconds
        self.update_next = now
        self.render_next = now
        self.apm_next = now
        self.stats_next = now

        # Do 1 iteration on loop anyways
        self.update()

        # Continue looping if game should run
        while Config.IS_RUNNING:
            self.update()

    def update(self):
        now = self.clock.elapsed_time.microseconds

        if now >= self.update_next:

            # Update all units
            for unit in self.units:
                if unit.removed:
                    continue

                unit.update()

            # Update players
            for player in self.players:
                player.update()

            # TODO score logging

            self.update_next += self.update_interval
            self.update_delta += 1
            self.ticks += 1

        if now >= self.apm_next:
            # Update algorithms for playe
            for player in self.players:
                if not player.algorithm:
                    continue

                player.algorithm.update()

            self.apm_next += self.apm_interval

        if self.doDisplay and now >= self.render_next:
            self.gui.update()
            self.gui.render()

            self.render_next += self.render_interval
            self.render_delta += 1

        if now >= self.stats_next:

            if self.doDisplay and self.doCaptionWindow:
                self.gui.caption()

            if self.doCaptionConsole:
                print("[FPS=%s, UPS=%s]" % (self.current_fps, self.current_ups))

            if True: # TODO add option
                for player in self.players:
                    player.apm = (player.apm + (player.apm_counter * 60)) / 2
                    player.apm_counter = 0

            self.current_fps = self.render_delta
            self.current_ups = self.update_delta
            self.render_delta = 0
            self.update_delta = 0
            self.stats_next += 1000000 # 1 second

File: cplusplus/src/Game.cpp
//
// Created by Per-Arne on 23.02.2017.
//

#include "Game.h"
#include "unit/UnitManager.h"


std::unordered_map<int, Game*> Game::games;

Game::Game(std::string map_file):
        config(Config::defaults()),
        map(map_file),
        state({map.MAP_WIDTH, map.MAP_HEIGHT, 10}), // Wait until map is loaded
        tilemap(map, *this) {
    init();
}

Game::Game(std::string map_file, Config config):
        config(config),
        map(map_file),
        state({map.MAP_WIDTH, map.MAP_HEIGHT, 10}), // Wait until map is loaded
        tilemap(map, *this){
    init();
}

void Game::init(){
    // State vector
    // 0 - Environment
    // 1 - Player ID

    // 1 - Unit/Building Type
    // 2 - Unit/Building Player
    // 3 - Unit/Building Health
    // Init environment

    players.reserve(Constants::MAX_PLAYERS);
    units.reserve(Constants::MAX_PLAYERS * Constants::MAX_UNITS);


    setMaxFPS(60);
    setMaxUPS(10);

    id = static_cast<uint8_t>(games.size());
    games[id] = this;

    // Update timings
    tick();
    timerInit();

    reset();

}





void Game::setMaxFPS(uint32_t fps_){
    max_fps = fps_;
    _render_interval = std::chrono::nanoseconds(1000000000 /  max_fps);


}

void Game::setMaxUPS(uint32_t ups_){
    max_ups = ups_;
    _update_interval = std::chrono::nanoseconds(1000000000 /  max_ups);


}

void Game::start(){
    timerInit();
    this->running = true;
}

void Game::stop(){
    this->running = false;
}

void Game::reset()
{


    // Reset all tiles
    for (auto &tile : tilemap.getTiles()) {
        tile.reset();
    }

	// Remove all units
	units.clear();

    // Reset all players
    for (auto &player : players) {
        player.reset();
        spawnPlayer(player);
    }

    // Reset tick counter
    ticks = 0;

    // Increase epsiode counter
    episode += 1;

    terminal = false;

    _reset();
    _onEpisodeStart();



}

void Game::update(){

    if (running && !terminal &&  now >= _update_next) {

        // Iterate through all units
        for(auto &unit : units) {
            if (unit.removedFromGame) continue;		// Skip unit that is removed from game
            unit.update();
        }

        // Iterate through all players
        for (auto &p : players) {
            p.update();
        }

        _update();

        // Update Counters and statistics
        _update_next += _update_interval;
        _update_delta += 1;
        ticks += 1;
    }

}


void Game::render(){
    if (now >= _render_next) {
        _render();


        _render_next += _render_interval;
        _render_delta += 1;
    }}

void Game::_render(){}
void Game::_caption(){}
void Game::_update(){}
void Game::_reset() {}


void Game::caption() {
    if (now >= this->_stats_next) {
        _caption();
        if (consoleCaptionEnabled) {
            std::cout << "[FPS=" << this->currentFPS << ", UPS=" << this->currentUPS << "]" << std::endl;
        }



        currentFPS = _render_delta;
        currentUPS = _update_delta;
        _render_delta = 0;
        _update_delta = 0;
        _stats_next += std::chrono::nanoseconds(1000000000);    // 1 Second
    }


}


void Game::tick() {
    now = std::chrono::high_resolution_clock::now();
}

void Game::timerInit() {
    tick();
    _render_next = now + _render_interval;
    _update_next = now + _update_interval;
    _stats_next = now + std::chrono::nanoseconds(0);
}







Game * Game::getGame(uint8_t id)
{
    Game *g = games.at(id);
    assert(g);
    return g;
}


bool Game::isTerminal(){

    // Skip check if terminal is already set
    if(terminal){
        return terminal;
    }

    int c = 0;
    for(auto &p : players) {
        if(p.isDefeated()){
            c++;
        }
    }

    bool isTerminal = (c == 1);

    terminal = isTerminal;

    _onEpisodeEnd();
    return terminal;
}


Player &Game::addPlayer() {
    players.emplace_back(*this, players.size());
    Player &player = players.back();

    spawnPlayer(player);
    return player;
}

void Game::spawnPlayer(Player &player) {
    // Retrieve spawn_point

    if(tilemap.spawnTiles.size() < players.size()){
        throw std::runtime_error(std::string("Failed to spawn player, There are not enough spawn tiles!"));
    }

    int spawnPointIdx = tilemap.spawnTiles[player.getId()];

    auto spawnTile = tilemap.tiles[spawnPointIdx];

    // Spawn Initial builder
    Unit &builder = player.spawn(spawnTile);

    // If auto-spawn town hall mechanic is activated
    if (config.instantTownHall) {
        // build Town-Hall
        builder.build(0);
    }

}

Unit & Game::getUnit(uint16_t idx)
{
    assert((idx >= 0 && idx < (units.size())) && "getUnit(idx) failed. Index not in range!");
    return units[idx];
}

size_t Game::getWidth() const{
    return map.MAP_WIDTH;
}

size_t Game::getHeight() const{
    return map.MAP_HEIGHT;
}


uint32_t Game::getMaxFPS() const {
    return max_fps;
}

uint32_t Game::getMaxUPS() const {
    return max_ups;
}

uint32_t Game::getFPS() const {
    return currentFPS;
}

uint32_t Game::getUPS() const {
    return currentUPS;
}

uint16_t Game::getEpisode() const {
    return episode;
}

uint64_t Game::getTicks() const{
    return this->ticks;
}

uint64_t Game::getGameDuration() const{
    return this->ticks / getTicksModifier();
}

int Game::getTicksModifier() const {
    return config.tickModifier;
}

uint8_t Game::getId() const {
    return id;
}

void Game::_onUnitCreate(Unit& unit) {
    //DEBUG("Unit Created: " + unit.name + " | " + unit.player_.getName() + "\n");
}


void Game::_onEpisodeStart() {
    //DEBUG("Episode " + std::to_string(episode) + " started.\n");
}

void Game::_onEpisodeEnd() {
    //DEBUG("Episode " + std::to_string(episode) + " ended.\n");
}

void Game::_onUnitDestroy(Unit& unit) {
    //DEBUG("Unit Destroyed: " + unit.name + " | " + unit.player_.getName() + "\n");
}



void Game::_onTileDeplete(Tile &){}





